Fix serialize plugin

tilersmyth opened the pull request:
Fixes #930. Conditional comparing functions always returned falsey. Utilizing toString() on the functions fixes the issue.

Gerrit0 commented:
I'm... wary about comparing stringified functions. This seems like a hack that corrects some other mistake in the serialization code. Why are the functions (instances?) different in the first place?

tilersmyth commented:
@Gerrit0 you could be right. This is the only fix I could come up with for the way the serializer currently adds plugins to the proper group. When the serializer `addComponent` function is executed it is comparing two distinct instances of the same function (object), which is why equality will never be true. 
Unless the serializer is refactored at a deeper level, using the .toString() method here might be adequate as we just want to check if it is the same function (it does not have to be same instance) to push it the right group for serialization.

tilersmyth commented:
@Gerrit0 you're right, this is not a great solution. After continuing to dig into the issue, it seems that the functions are not equal because the serialization plugin is referencing the class declaration where as internally the the `serializeGroupSymbol` is referencing the class (due to it being referenced prior to being compiled - I think). See output comparison below.. 

For my purposes of getting the serializer to work, I have now simply updated the `match` function to look for equality in `serializeGroupSymbol` names with the following: 

```
match = Array.from(this.router.values()).find(v => v.symbol.name === component.serializeGroupSymbol.name)
```
This has been working well, but let me know if you think this solution is worth a PR. Otherwise, I will just leave it alone until a better fix comes along. Thanks for your input on this. 


## Current serializeGroupSymbol output (Plugin vs. Internal)
**Plugin:** 
```
function Reflection(parent, name, kind) {
        this.name = '';
        this.flags = [];
        this.id = REFLECTION_ID++;
        this.parent = parent;
        this.name = name;
        this.originalName = name;
        this.kind = kind;
    }
```

**Internal:**
```
class Reflection {
    constructor(name, kind, parent) {
        this.name = '';
        this.flags = new ReflectionFlags();
        this.id = REFLECTION_ID++;
        this.parent = parent;
        this.name = name;
        this.originalName = name;
        this.kind = kind;
    }
    kindOf(kind) {
        const kindArray = Array.isArray(kind) ? kind : [kind];
        return kindArray.some(kind => (this.kind & kind) !== 0);
    }
    getFullName(separator = '.') {
        if (this.parent && !this.parent.isProject()) {
            return this.parent.getFullName(separator) + separator + this.name;
        }
        else {
            return this.name;
        }
    }
    setFlag(flag, value = true) {
        this.flags.setFlag(flag, value);
    }
    getAlias() {
        if (!this._alias) {
            let alias = this.name.replace(/[^a-z0-9]/gi, '_').toLowerCase();
            if (alias === '') {
                alias = 'reflection-' + this.id;
            }
            let target = this;
            while (target.parent && !target.parent.isProject() && !target.hasOwnDocument) {
                target = target.parent;
            }
            if (!target._aliases) {
                target._aliases = [];
            }
            let suffix = '', index = 0;
            while (target._aliases.indexOf(alias + suffix) !== -1) {
                suffix = '-' + (++index).toString();
            }
            alias += suffix;
            target._aliases.push(alias);
            this._alias = alias;
        }
        return this._alias;
    }
    hasComment() {
        return this.comment ? this.comment.hasVisibleComponent() : false;
    }
    hasGetterOrSetter() {
        return false;
    }
    getChildByName(arg) {
        const names = Array.isArray(arg) ? arg : arg.split('.');
        const name = names[0];
        let result;
        this.traverse((child) => {
            if (child.name === name) {
                if (names.length <= 1) {
                    result = child;
                }
                else if (child) {
                    result = child.getChildByName(names.slice(1));
                }
            }
        });
        return result;
    }
    isProject() {
        return false;
    }
    findReflectionByName(arg) {
        const names = Array.isArray(arg) ? arg : arg.split('.');
        const reflection = this.getChildByName(names);
        if (reflection) {
            return reflection;
        }
        else if (this.parent) {
            return this.parent.findReflectionByName(names);
        }
    }
    traverse(callback) { }
    toObject() {
        const result = {
            id: this.id,
            name: this.name,
            kind: this.kind,
            kindString: this.kindString,
            flags: {}
        };
        if (this.originalName !== this.name) {
            result.originalName = this.originalName;
        }
        if (this.comment) {
            result.comment = this.comment.toObject();
        }
        Object.getOwnPropertyNames(ReflectionFlags.prototype).forEach(name => {
            const descriptor = Object.getOwnPropertyDescriptor(ReflectionFlags.prototype, name);
            if (typeof descriptor.get === 'function' && this.flags[name] === true) {
                result.flags[name] = true;
            }
        });
        if (this.decorates) {
            result.decorates = this.decorates.map((type) => type.toObject());
        }
        if (this.decorators) {
            result.decorators = this.decorators.map((decorator) => {
                const result = { name: decorator.name };
                if (decorator.type) {
                    result.type = decorator.type.toObject();
                }
                if (decorator.arguments) {
                    result.arguments = decorator.arguments;
                }
                return result;
            });
        }
        this.traverse((child, property) => {
            if (property === TraverseProperty.TypeLiteral) {
                return;
            }
            let name = TraverseProperty[property];
            name = name.substr(0, 1).toLowerCase() + name.substr(1);
            if (!result[name]) {
                result[name] = [];
            }
            result[name].push(child.toObject());
        });
        return result;
    }
    toString() {
        return ReflectionKind[this.kind] + ' ' + this.name;
    }
    toStringHierarchy(indent = '') {
        const lines = [indent + this.toString()];
        indent += '  ';
        this.traverse((child, property) => {
            lines.push(child.toStringHierarchy(indent));
        });
        return lines.join('\n');
    }
}
```

Gerrit0 commented:
Thanks for looking in to this more! Reviewing the serialization has been on my list for a while and this should be helpful in finding what should be done to fix this. Checking names is better than `.toString`, but I'd still like to resolve the root issue.

tilersmyth commented:
Ok cool! I will be following.
